src/world/map.c

```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
+			quadrant = 0;
+		}
+	}
+
+	// Check if sub tile height is any different compared to actual surface tile height
+	int x2 = x;
+	int y2 = y;
+	if(scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG_FULL_TILE){
+		x2 += 16;
+		y2 += 16;
+	}else{
+		x2 += ScenerySubTileOffsets[quadrant & 3].x - 1;
+		y2 += ScenerySubTileOffsets[quadrant & 3].y - 1;
+	}
+	base_height = map_element_height(x2, y2);
+	// If on water
+	if(base_height & 0xFFFF0000){
+		// base_height2 is now the water height
+		base_height >>= 16;
+		if(targetHeight == 0){
+			isOnWater = true;
+		}
+	}
+	if(targetHeight == 0){
+		targetHeight = base_height;
+	}
+
+	if(!(gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) && 
+		!gCheatsSandboxMode && 
+		!map_is_location_owned(x, y, targetHeight)){
+
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if(flags & GAME_COMMAND_FLAG_APPLY && !(flags & 0x40)){
+		footpath_remove_litter(x, y, targetHeight);
+		if(!gCheatsDisableClearanceChecks && (scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG_ALLOW_WALLS)) {
+			map_remove_walls_at(x, y, targetHeight, targetHeight + scenery_entry->small_scenery.height);
+		}
+	}
+
+	rct_map_element* surface_element = map_get_surface_element_at(x / 32, y / 32);
+
+	if(!gCheatsDisableClearanceChecks && (surface_element->properties.surface.terrain & 0x1F)){
+		int water_height = ((surface_element->properties.surface.terrain & 0x1F) * 16) - 1;
+		if(water_height > targetHeight){
+			gGameCommandErrorText = STR_CANT_BUILD_THIS_UNDERWATER;
+			*ebx = MONEY32_UNDEFINED;
+			return;
+		}
+	}
+
+	if(!gCheatsDisableClearanceChecks && !(scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG18)){
+		if(isOnWater){
+			gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_ON_LAND;
+			*ebx = MONEY32_UNDEFINED;
+			return;
+		}
+
+		if(surface_element->properties.surface.terrain & 0x1F){
+			if(((surface_element->properties.surface.terrain & 0x1F) * 16) > targetHeight){
+				gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_ON_LAND;
+				*ebx = MONEY32_UNDEFINED;
+				return;
+			}
+		}
+	}
+
+	if (!gCheatsDisableClearanceChecks &&
+		(scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG_REQUIRE_FLAT_SURFACE) &&
+		!supportsRequired &&
+		!isOnWater &&
+		(surface_element->properties.surface.slope & 0x1F)) {
+
+		gGameCommandErrorText = STR_LEVEL_LAND_REQUIRED;
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if (!gCheatsDisableSupportLimits && 
+		!(scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG18) && 
+		supportsRequired) {
+
+		if(!isOnWater){
+			if((surface_element->properties.surface.terrain & 0x1F) || 
+				(surface_element->base_height * 8) != targetHeight){
+
+				gGameCommandErrorText = STR_LEVEL_LAND_REQUIRED;
+				*ebx = MONEY32_UNDEFINED;
+				return;
+			}
+
+		}else{
+			gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_ON_LAND;
+			*ebx = MONEY32_UNDEFINED;
+			return;
+		}
+	}
+
+	int bp = quadrant;
+	int zLow = targetHeight / 8;
+	int zHigh = zLow + ((scenery_entry->small_scenery.height + 7) / 8);
```

Comment:
`ceil2(scenery_entry->small_scenery.height, 8) / 8` might make this a bit more obvious?